### ISSUE TITLE
Big vs Little Endianness

### ISSUE THREAD

**Issue by llamafilm** [Cyan4973/xxHash#45]
I'm attempting to use this tool to verify checksums created by Pomfort Silverstack.  They never match, and the Silverstack developers say this is because they use little endian, and xxhsum uses big endian.  Would it make sense to add this switch as a command-line option?  Or at least specify in the documentation which version is being used, so that users aren't surprised when hashes don't match.

**Comment by Cyan4973** [comment]
Good point.
I believe it makes sense is to specify which display convention is used by `xxhsum`.

**Comment by llamafilm** [comment]
Is it just a display convention, or is it a different algorithm?  I'm wondering if it's possible to convert between the two.  For example, one particular file shows 706e68245bac7ad8 from xxhsum and d87aac5b24686e70 from Silverstack.

**Comment by Cyan4973** [comment]
Yes, it's only a display convention.
Both numbers are equal if you consider endianess : 70 6e 68 ... <--> ... 68 6e 70

**Comment by Cyan4973** [comment]
Display convention is now mentioned in source code, although it's might not be the place where you would like to find it.

Where does it make most sense to specify it ?

**Comment by llamafilm** [comment]
Thanks for your quick response!  I suggest adding a command-line argument, and show the syntax in the help text, for example:
-e# : display endianness : 0=little, 1=big (default 1)

Otherwise, a line in the readme would make sense.

**Comment by Cyan4973** [comment]
There's a new command line option within [dev branch](https://github.com/Cyan4973/xxHash/tree/dev)
which purpose is to allow optional selection of little endian convention.

New command is `--little-endian`, and is documented using `./xxhsum -h`.

For your testing

**Comment by t-mat** [comment]
Is there any chance to expose "canonical" digest function, something like this:

```
XXH_errorcode XXH32_canonicalDigest (const XXH32_state_t* statePtr, unsigned char* digestBuf,
      size_t digestBufSizeInBytes);
```

> - statePtr : XXH32 state
> - digestBuf : Output buffer for canonical XXH32 value.
> - digestBufSizeInBytes : Size of digestBuf in bytes. At least 4 bytes for XXH32.
>
> XXH32_canonicalDigest() writes canonical XXH32 digest value as byte array to digestBuf.
> (I suppose there are many objections for prototype and arguments.)

Main purpose of this function is decreasing freedom of presentation.  Also this prototype could provide similar functions for current and future XXH families.
Perhaps it's good chance to prevent future confusion, if we recommend such function for human readable and string type presentation such as JSON (currently, JavaScript doesn't have 64bit integer).

**Comment by Cyan4973** [comment]
Pretty good point @t-mat,
this is indeed a good way to make sure everyone gets the same representation,
and it's indeed a good prototype to add new members with wider hashes.

The only downside is that it's likely a bit slower than using direct integer representation.
But well, as long as it is an additional option, I see no problem with that.

**Comment by Cyan4973** [comment]
In b5e9b75d84313b3409c54b42d3727fea76a3c735, there is a first attempt at providing `_canonicalDigest()` functions, as suggest by @t-mat.

3 minors things I believe could be reviewed before reaching master : 

- Prototype requires size of `buffer`.
  Is that really useful ?
  The amount of bytes to be written into `buffer` is pretty clear : 4 for 32 bits, 8 for 64 bits.
  Requiring the buffer size allow an extra check at the cost of an heavier interface.
  For comparison, `md5sum` interface doesn't require it.

- The final write operation, using big-endian convention, is not optimized.

- How to test it ?

**Comment by t-mat** [comment]
> Prototype requires size of buffer.
> Is that really useful ?

Since xxHash is minimal library, heavier interface may be over engineering.  But here I dare to write "may be good" part of providing size of buffer.

 - I think, we may be able to justify `XXHnn_canonicalDigest()`'s relatively heavy looking and processing.
    - Because it's highly related to relatively slow output procedure for interexchange, such as pipe/file/network.
    - High performance applications may prefer to use `XXHnn_digest()` directly.
 - It may be good for FFI (JNI, P/Invoke, etc).
 - It prevents stupid migration error, something like this:

```C
struct Buf {
   char digest[4];
   char filename[32];
   // ...;
};

Buf buf;

// before migration
void f_old() {
    assert(XXH_OK == XXH32_canonicalDigest(xxh, buf.digest, sizeof(buf.digest)));
    assert(XXH_OK == XXH32_canonicalDigest(xxh, buf.digest));
}

// after migration.  Suppose I fogot to change buf.digest[].
void f_new() {
    // Runtime error
    assert(XXH_OK == XXH64_canonicalDigest(xxh, buf.digest, sizeof(buf.digest)));

    // Overwrites buf.filename[] silently.
    // Of course, this is programmer's error.
    assert(XXH_OK == XXH64_canonicalDigest(xxh, buf.digest));
}
```

Another possible interface is type based conversion functions.  In this case, we can decouple canonical representation functions and `XXHnn_state_t`.  Since type relation graph becomes `XXHnn_state_t` --> `XXHnn_hash_t` <--> `XXHnn_canonical_t`, it's more composable.  For example, we can make canonical representation from `XXHnn()` hash value.

```C
// xxhash.h
typedef unsigned int XXH32_hash_t;  // Hash value
typedef char[4] XXH32_canonical_t;  // Canonical representation (struct/union is also good candidate)

// uint32_t to uint8_t[4]
void XXH32_hashToCanonical(XXH32_hash_t src, XXH32_canonical_t dst);

// uint8_t[4] to uint32_t
XXH32_hash_t XXH32_canonicalToHash(const XXH32_canonical_t src);

// application
void f(const void* dataPtr, size_t dataBytes) {
    XXH32_canonical_t canonical;
    XXH32_hash_t xxh = XXH32(dataPtr, dataBytes, 0);
    XXH32_hashToCanonical(xxh, canonical);
    printf("%02x%02x%02x%02x\n", canonical[0], canonical[1], canonical[2], canonical[3]);
}
```

> The final write operation, using big-endian convention, is not optimized.

Hmm, `XXH_swapNN() + memcpy()`?

> How to test it ?

If we can change xxhsum, the following change seems good enough for me:

   - Rewrite `BMK_display_BigEndian()` to use canonical functions.
   - Add test target file `TEST_FILE` to repository, and `xxhsum TEST_FILE | grep "*EXPECTED_HASH_VALUE*  TEST_FILE"` to Makefile's `test` target.
       - Or if we can add random file generator named xxh_testgen, `xxh_testgen --seed SEED | xxhsum - | ...`

If we can add dedicated sanity checking mode (`xxhsum --test-sanity`),

   - Prebuild PRNG seed and expected XXH32/XX64 hash value and canonical hash byte array pair tables.  To avoid endianness magic, we should provide both byte array and integer hash value.
       - Make test memory image by PRNG with given seed (`lz4/programs/datagen.c` seems good).
       - Generate hash value and canonical hash byte array of above image.
       - Compare results with given hash value and canonical byte array.

**Comment by Cyan4973** [comment]
> Another possible interface is type based conversion functions.

I like it because it avoids being in "competition" with normal digest operation, and just complements it.

> typedef unsigned int XXH32_hash_t;  // Hash value

I'm wondering if such a typedef really helps, since it masks the fact that `XXH32_hash_t` is `unsigned int`. I suspect there are benefits using `unsigned int` directly, such as `printf` formatting and simple arithmetic. I understand it's also possible with `XXH32_hash_t`, it's just a bit weird to do such operations on this type.

Even better would be `uint32_t` ... still this issue of `<stdint.h>` C99 support which is not universal ...

**Comment by t-mat** [comment]
note: I don't blame Pomfort developers.  Below is general discussion.

> I like it because it avoids being in "competition" with normal digest operation, and just complements it.

Agreed.  Since it becomes one function for one thing, it is better than previous one for testing and future maintenance.  Sorry for my hasty.  

> I'm wondering if such a typedef really helps, since it masks the fact that XXH32_hash_t is unsigned int.

tl;dr - It's strong type fanboy's personal preference!

Usually, from application programmer's point of view, unique type is better for type safety.  In this case, unfortunately, since C (and C++) doesn't provide straightfoward strong integer typedef, it's just a cosmetic and conventional typedef. (Technically, C++ can do it [hacky way](http://ideone.com/I18g9K))
But logically, application programmer should not compare XXH32 and CRC32 even both of them are uint32_t.  Also they should not increment or apply any arithmetic/binary logic operation for XXH32_hash_t without internal knowledge.
When they want to do arithmetic/binary logic operations for hash key or something, they must recognize it's already not a valid XXH32 hash value after arithmetic operation like `int*` to `uintptr_t` conversion.

I think for 80% of applications, only relational operators (especially `==`) and value generators (XXH functions) are needed to implement their application.

> I suspect there are benefits using unsigned int directly, such as printf formatting and simple arithmetic.

Right.  But since this issue itself shows problem of direct access and printf, it would be nice to avoid it for representing hash value for end users and interexchanging.  Although it's quite handy for internal usage and debugging.

So my point is, since xxHash follows basic "Spirit of C", there is absolutely no problem to expose raw type.  And I also believe xxHash also follows "code as document" principle, there is little benefit to declare unique type which indicates: Don't touch or inspect it without enough knowledge.  Treat it as mysterious `unsigned int` size value.

**Comment by Cyan4973** [comment]
> I think for 80% of applications, only relational operators (especially ==) and value generators (XXH functions) are needed to implement their application.

All right. So now let's assume  xxhash prototypes are modified so that : 
```
typedef unsigned int XXH32_hash_t;
XXH_PUBLIC_API XXH32_hash_t  XXH32_digest (const XXH32_state_t* statePtr);
```

What should be the type of `seed` then ?
```
XXH_PUBLIC_API XXH_errorcode XXH32_reset  (XXH32_state_t* statePtr, unsigned int seed);
```

Should it be XXH32_hash_t or should it remain `unsigned int` ? (even if it's just cosmetic)

**Comment by Cyan4973** [comment]
I've just realized that you are also proposing a strong type for canonical representation : 

```
typedef struct { char digest[4]; } XXH32_canonical_t;  // Canonical representation
```

I suspect this can be a problem to write the result into an arbitrary buffer. Strict aliasing will get in the way/

**Comment by t-mat** [comment]
> What should be the type of seed then ?

Since seed is not a hash value and it's required to be just a 32bit integer, `unsigned int` (or `uint32_t`) is best type.  We should avoid to use `XXH32_hash_t` for the seed.  Because purpose of `XXH32_hash_t` is closing the circle of operations.
Only XXH32 related hash functions generate hash value `XXH32_hash_t`, and canonical functions convert it to/from canonical form.  If we use `XXH32_hash_t` for the seed, users generate it by their hand (e.g. `XXH32_hash_t mySeed = 0; XXH32_reset(xxhPtr, mySeed);`).  And if user generates `XXH32_hash_t` by hand, circle of operations will be broken.

> I suspect this can be a problem to write the result into an arbitrary buffer. Strict aliasing will get in the way.

Right.  I suppose you are worrying about efficiency of struct based implementation and considering array-like definition `typedef char XXH32_canonical_t[4];`.
I think array is fine.  It simplifies everything.  Only one problem is it looks like value when we pass it to the another function and many people don't like it.  But in this case, it is matter of preference.

For the performance problem, I'm sure about the following `serialize()` generates optimal code with gcc/clang/MSVC.

```
typedef unsigned int XXH32_hash_t;
typedef struct { char digest[4]; } XXH32_canonical_t;

size_t serialize(char* d, XXH32_hash_t xxh) {
    XXH32_canonical_t c;
    XXH32_hashToCanonical(xxh, &c);
    memcpy(d, &c.digest, sizeof(c.digest));
    return sizeof(c.digest);
}

size_t deserialize(const char* s, XXH32_hash_t* xxh) {
    XXH32_canonical_t c;
    memcpy(&c.digest, s, sizeof(c.digest));
    *xxh = XXH32_canonicalToHash(&c);
    return sizeof(c.digest);
}
```

If LTO is enabled, their optimizer [removes memcpy() and entire function call](https://goo.gl/OKqeqY).
I'm not sure about deserialize() pattern, so I need to study about it, but it seems good enough for those compilers.
(I also wrote [complete example](https://gist.github.com/t-mat/b0a90319607d23b72828) for further study.)

For dynamic link case, I believe array is always faster.  If struct definition is needed to be optimized, we should define another special de/serialization function something like this:`void XXH32_serializeCanonical(char* dst, XXH32_hash_t hash);`.

**Comment by Cyan4973** [comment]
> Since seed is not a hash value and it's required to be just a 32bit integer, unsigned int (or uint32_t) is best type. 

Good

>  I suppose you are worrying about efficiency of struct based implementation and considering array-like definition `typedef char XXH32_canonical_t[4];` .

Not really, although it's a good thing you checked that outcome with very convincing tests.

My worry was really about strict-aliasing, and its potential consequences, when a memory area is being accessed by 2 different pointer types (risks of one operation being "optimized" directly into register, with the outcome made not accessible to other pointers).

One way to check for such a risk is to use `-Wstrict-aliasing=1`, which is more stringent than `-Wstrict-aliasing` alone (it defaults to 3).

Your godbolt example compiles fine with this setting, and I made a few variations of this test, which look ok too. Maybe it's related to the fact that underlying type is still base on `char*`. Anyway, thanks to your tests, I'm less worried now.

#### Minor modifications : 

I'm considering the following modifications to your suggestion : 

```
typedef struct { unsigned char digest[4]; } XXH32_canonical_t;  // Canonical representation
```
(note the `unsigned`)

```
void XXH32_CanonicalFromHash(XXH32_canonical_t* dst, XXH32_hash_t hash);
```
(note the reverse order of arguments, to feature "destination first" design.
Unfortunately, I believe it would then be confusing to keep initial name `HashToCanonical`,
so function name is modified too)

Is `XXH32_hash_t XXH32_canonicalToHash(const XXH32_canonical_t* src);` really useful ?

**Comment by t-mat** [comment]
> My worry was really about strict-aliasing, and its potential consequences

I see.  I'm not 100% sure about standard, but I think it's safe.  Because [C99 (N1256): 6.7.2.1/13](http://port70.net/~nsz/c/c99/n1256.html#6.7.2.1p13) says :

> Within a structure object, the non-bit-field members and the units in which bit-fields 
> reside have addresses that increase in the order in which they are declared. 
> A pointer to a structure object, suitably converted, points to its initial member 
> (or if that member is a bit-field, then to the unit in which it resides), and vice versa. 
> There may be unnamed padding within a structure object, but not at its beginning.

Since first member is `char[]`, we can say it is "suitably converted" to/from `char*`.  Therefore strict aliasing isn't violated.

> (note the unsigned)
> "destination first" design

:+1: :+1:

> Is XXH32_hash_t XXH32_canonicalToHash(const XXH32_canonical_t* src); really useful ?

If we implement canonical form, it's absolutely useful.  For example, suppose we have some JSON for file consistency checking:

```
{
  "my-file": "01234567",
  "another-file": "89abcdef"
}
```

We can transform it by the following sequence:

 - char[]: `"01234567"` (Canonical hex-string representation)
 - unsigned char[]: `{ 0x01, 0x23, 0x45, 0x67 }` (Canonical unsigned char array)
 - `XXH32_canonical_t` : (Canonical struct)
 - `XXH32_hash_t` : (Native hash value)

In this list, latter one is easier to handle in the code, former one is easier to recognize and interexchange.  And I believe almost all programmers prefer to use `XXH32_hash_t` in their internal code for efficiency and readability (even in C, they can use `==` freely!).
Since xxHash is hash function library, hash value (`XXH32_hash_t`) centric design is natural choice.  So it's usually useful to keep this list's conversion sequence for both direction (former to latter, and vice versa).

**Comment by Cyan4973** [comment]
OK, consider me convinced.

Here is the API I'm preparing following your suggestions : 
```
/* **************************
*  Canonical representation
****************************/
typedef struct { unsigned char digest[4]; } XXH32_canonical_t;
typedef struct { unsigned char digest[8]; } XXH64_canonical_t;

XXH_PUBLIC_API void XXH32_canonicalFromHash(XXH32_canonical_t* dst, XXH32_hash_t hash);
XXH_PUBLIC_API void XXH64_canonicalFromHash(XXH64_canonical_t* dst, XXH64_hash_t hash);

XXH_PUBLIC_API XXH32_hash_t XXH32_hashFromCanonical(const XXH32_canonical_t* src);
XXH_PUBLIC_API XXH64_hash_t XXH64_hashFromCanonical(const XXH64_canonical_t* src);

/*! Default result types for XXH functions are primitive unsigned 32 and 64 bits.
*   The canonical representation uses human-readable write convention, aka big-endian (large digits first).
*   These functions allow transformation of hash result into and from its canonical format.
*   This way, hash values can be written into a file / memory, and remain comparable with different systems and programs.
*/
```

Now my last question is about the canonical format itself.

Up to now, I've been using big-endian, because it is a direct human-readable representation of a U32 or U64 value. Debugger will read it this way for example, or `printf` will write it this way.

All this sounds good as long as the objective is merely to display some values, since the effort to display anything even in text mode is way heavier than an endian conversion.

Thinking again, I'm now wondering if canonical format would be used for inter-operable storage and automated comparisons. In which case, it's no longer about displaying a few values, but converting and comparing a large number of them.
I'm wondering if the endian conversion could become a contributor to this chain, or if it remains a totally negligible cost. (for little-endian platforms, which are likely the most common users of xxHash)

**Comment by Cyan4973** [comment]
Finally decided to move on : a8812895215c17d347f1a8c4cd9dbabe04e0f2f6

The final interface is the one proposed in earlier post, using big endian convention.
I believe the cost of endian conversion will remain negligible in most if not all circumstances.

Also : 
- optimized endian conversion, using existing XXH assets (endian detection and conversion)
- xxhsum uses canonical API to display file hash results

**Comment by Cyan4973** [comment]
Merged into master

**Comment by t-mat** [comment]
> we are in fact talking about a storage convention
> As a counter example, lz4 uses xxh32 checksum stored as little-endian.

LZ4 checksum is unfortunate exception, but since checksum is not directly exposed to user and LZ4 defines every aspect of interoperability, I think direct represantation of `XXH32_hash_t` is valid usage.
IMO canonical represantation and its APIs are good for:

 - Applications which mainly consider about interoperability and consistency.
 - Closed system (which don't exchange xxHash value with 3rd party systems) developers who don't want to develop and maintain entire interoperable definitions and functions.
 - Lazy people :sunglasses:

It doesn't cover 100% of applications, but it's reasonably good enough for many of them.

> I believe the cost of endian conversion will remain negligible in most if not all circumstances.

I agree.  If it becomes real problem, they should consider to use `XXHnn_hash_t` directly.

> Merged into master

I'm late.  But just in case, extra static assert before `memcpy()` would be nice to indicate "we are taking care of it"

```
XXH_STATIC_ASSERT(sizeof(XXH32_canonical_t) == sizeof(XXH32_hash_t));
```

**Comment by Cyan4973** [comment]
Good point !
84e6218cd5bc501f7e79b36274383bf97957e830
